File: Utilities/Others/Input.m
function varargout = Input(varargin)
% Process the input of algorithm design.

%----------------------------Copyright-------------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------

switch varargin{end}
    case 'data'
        value   = varargin{1};
        Setting = varargin{2};
        % input problem
        if any(strcmp(value,'Problem'))
            ind = find(strcmp(value,'Problem'));
            prob  = value{ind+1};
            varargout{1} = prob;
        else
            error('Please set the targeted problem.');
        end
        if strcmp(Setting.Mode,'design') && any(strcmp(value,'InstanceTrain')) && any(strcmp(value,'InstanceTest'))
            ind = find(strcmp(value,'InstanceTrain'));
            instanceTrain = value{ind+1}; % indexes of problem instances for designing algorithm
            ind = find(strcmp(value,'InstanceTest'));
            instanceTest  = value{ind+1}; % indexes of problem instances for testing the designed algorithm
            varargout{2} = instanceTrain;
            varargout{3} = instanceTest;
        elseif strcmp(Setting.Mode,'solve') && any(strcmp(value,'InstanceSolve'))
            ind = find(strcmp(value,'InstanceSolve'));
            instanceSolve = value{ind+1}; % indexes of problem instances to be solved
            varargout{2} = instanceSolve;
        else
            error('Please set the targeted problem instance indexes.');
        end

    case 'parameter'
        value   = varargin{1};
        Setting = varargin{2};

        if any(strcmp(value,'AlgP'))
            ind = find(strcmp(value,'AlgP'));
            Setting.AlgP = value{ind+1};
        end
        if any(strcmp(value,'AlgQ'))
            ind = find(strcmp(value,'AlgQ'));
            Setting.AlgQ = value{ind+1};
        end
        if any(strcmp(value,'Archive'))
            ind = find(strcmp(value,'Archive'));
            Setting.Archive = value{ind+1};
        end
        if any(strcmp(value,'LSRange'))
            ind = find(strcmp(value,'LSRange'));
            Setting.LSRange = value{ind+1};
        end
        if any(strcmp(value,'IncRate'))
            ind = find(strcmp(value,'IncRate'));
            Setting.IncRate = value{ind+1};
        end
        if any(strcmp(value,'ProbN'))
            ind = find(strcmp(value,'ProbN'));
            Setting.ProbN = value{ind+1};
        end
        if any(strcmp(value,'ProbFE'))
            ind = find(strcmp(value,'ProbFE'));
            Setting.ProbFE = value{ind+1};
        end
        if any(strcmp(value,'InnerFE'))
            ind = find(strcmp(value,'InnerFE'));
            Setting.InnerFE = value{ind+1};
        end
        if any(strcmp(value,'AlgN'))
            ind = find(strcmp(value,'AlgN'));
            Setting.AlgN = value{ind+1};
        end
        if any(strcmp(value,'AlgFE'))
            ind = find(strcmp(value,'AlgFE'));
            Setting.AlgFE = value{ind+1};
        end
        if any(strcmp(value,'AlgRuns'))
            ind = find(strcmp(value,'AlgRuns'));
            Setting.AlgRuns = value{ind+1};
        end
        if any(strcmp(value,'Metric'))
            ind = find(strcmp(value,'Metric'));
            Setting.Metric = value{ind+1};
        end
        if any(strcmp(value,'Compare'))
            ind = find(strcmp(value,'Compare'));
            Setting.Compare = value{ind+1};
        end
        if any(strcmp(value,'Evaluate'))
            ind = find(strcmp(value,'Evaluate'));
            Setting.Evaluate = value{ind+1};
        end
        if any(strcmp(value,'Tmax'))
            ind = find(strcmp(value,'Tmax'));
            Setting.Tmax = value{ind+1};
        end
        if any(strcmp(value,'Thres'))
            ind = find(strcmp(value,'Thres'));
            Setting.Thres = value{ind+1};
        end
        if any(strcmp(value,'RacingK'))
            ind = find(strcmp(value,'RacingK'));
            Setting.RacingK = value{ind+1};
        end
        if any(strcmp(value,'Surro'))
            ind = find(strcmp(value,'Surro'));
            Setting.Surro = value{ind+1};
        end
        if any(strcmp(value,'AlgFile'))
            ind = find(strcmp(value,'AlgFile'));
            Setting.AlgFile = value{ind+1};
        end
        if any(strcmp(value,'AlgName'))
            ind = find(strcmp(value,'AlgName'));
            Setting.AlgName = value{ind+1};
        end
        varargout{1} = Setting;

    case 'check'
        Setting = varargin{1};
        switch Setting.Mode
            case 'design'
                % AlgP or AlgQ should be equal to 1
                if Setting.AlgP > 1 && Setting.AlgQ > 1
                    error(['Setting.AlgP or Setting.AlgQ should be equal to 1 - ' ...
                        'For algorithms with multiple search pathways (AlgP>1), ' ...
                        'each search pathway should have only one search operator (AlgQ=1). ' ...
                        'For algorithms with a single search pathway (AlgP=1), ' ...
                        'the search pathway can have multiple search operators (AlgQ>=1).'])
                end

                % set Tmax and Thres when using runtimeFE or runtimeSec as the performance metric
                if strcmp(Setting.Metric,'runtimeFE') && isempty(Setting.Tmax)
                    Setting.Tmax = Setting.ProbFE;
                end
                if strcmp(Setting.Metric,'runtimeFE') && isempty(Setting.Thres)
                    error(['Please set "Setting.Thres" as the lowest acceptable performance ' ...
                        'of the design algorithms, the performance can be the solution quality.'])
                end
                if strcmp(Setting.Metric,'runtimeSec') && isempty(Setting.Tmax)
                    error('Please set "Setting.Tmax" as the maximum runtime (seconds).')
                end
                if strcmp(Setting.Metric,'runtimeSec') && isempty(Setting.Thres)
                    error(['Please set "Setting.Thres" as the lowest acceptable performance ' ...
                        'of the design algorithms, the performance can be the solution quality.'])
                end

                %  time (Tmax) should be function evaluations when using auc
                if strcmp(Setting.Metric,'auc') && numel(Setting.Tmax) <= 1
                    error(['"Setting.Tmax" should contain multiple time points. The time ' ...
                        'points should the numbers of function evaluations spent during ' ...
                        'the alorithm execution.'])
                end

                % thresholds should be corresponding to time points when using auc
                if strcmp(Setting.Metric,'auc') && numel(Setting.Thres) ~= numel(Setting.Tmax)
                    error(['The number of thresholds in "Setting.Thres" should be equal to ' ...
                        'the number of time points in "Setting.Tmax". "Setting.Thres" ' ...
                        'refers to the lowest acceptable performance of the design' ...
                        ' algorithms, the performance can be the solution quality.'])
                end

                % the "racing" evaluation method should be used with the "statictic" algorithm comparing method
                if strcmp(Setting.Evaluate,'racing') && ~strcmp(Setting.Compare,'statistic')
                    error(['The "racing" evaluation method should be used with the ' ...
                        'algorithm comparing method of "statistic". '])
                end

                % should set RacingK when using the "racing" evaluation method
                if strcmp(Setting.Evaluate,'racing') && isempty(Setting.RacingK)
                    error(['Please set "Setting.K" as the number of instances evaluated ' ...
                        'before the first round of racing.'])
                end

                % should set Surro when using the "approximate" evaluation method
                if strcmp(Setting.Evaluate,'approximate') && isempty(Setting.Surro)
                    error(['Please set "Setting.Surro" as the number of exact performance evaluations' ...
                        ' when using surrogate.'])
                end

                % it is not necessary to use "statistic" algorithm comparing method when using the "approximate" evaluation method
                if strcmp(Setting.Evaluate,'approximate') && strcmp(Setting.Compare,'statistic')
                    error(['It is not necessary to use the "statistic" algorithm ' ...
                        'comparing method when using the "approximate" evaluation method.'])
                end

                % should run the design multiple times when using the "statistic" comparsion method
                if strcmp(Setting.Compare,'statistic') && Setting.AlgRuns == 1
                    error(['Please run the design multiple times (Setting.AlgRuns>1)' ...
                        ' when using the "statistic" comparsion method.'])
                end

                % better to have a large population size when involving the EDA operator
                if Setting.ProbN < 5 && Setting.AlgP > 1
                    warning(['It is better to have a large population size if ' ...
                        'involving the EDA operator'])
                end

                % AlgP cannot be very large
                if Setting.AlgQ > 4
                    warning(['AlgQ is recommended to be larger than 4 for ' ...
                        'discrete and permutation problems due to the lack' ...
                        ' of so many search operators'])
                end

                % better to have a large number of training intrances or have a large number of algorithm runs, in order to make the statistical test discriminative
                if strcmp(Setting.Compare,'statistic')
                    warning(['It is better to have a large number of training intrances ' ...
                        'or have a large number of algorithm runs (set AlgRun to a large number), ' ...
                        'in order to make the statistical test discriminative.']);
                end

            case 'solve'
                % should specify the file or name of the algorithm being used
                if isempty(Setting.AlgFile) && isempty(Setting.AlgName)
                    error(['Please specify an algorithm file in Setting.AlgFile or ' ...
                        'specify an algorithm name in Setting.AlgName.'])
                end

                % set Tmax and Thres when using runtimeFE or runtimeSec as the performance metric
                if strcmp(Setting.Metric,'runtimeFE') && isempty(Setting.Tmax)
                    Setting.Tmax = Setting.ProbFE;
                end
                if strcmp(Setting.Metric,'runtimeFE') && isempty(Setting.Thres)
                    error(['Please set "Setting.Thres" as the lowest acceptable performance ' ...
                        'of the design algorithms, the performance can be the solution quality.'])
                end
                if strcmp(Setting.Metric,'runtimeSec') && isempty(Setting.Tmax)
                    error('Please set "Setting.Tmax" as the maximum runtime (seconds).')
                end
                if strcmp(Setting.Metric,'runtimeSec') && isempty(Setting.Thres)
                    error(['Please set "Setting.Thres" as the lowest acceptable performance ' ...
                        'of the design algorithms, the performance can be the solution quality.'])
                end

                %  time (Tmax) should be function evaluations when using auc
                if strcmp(Setting.Metric,'auc') && numel(Setting.Tmax) <= 1
                    error(['"Setting.Tmax" should contain multiple time points. The time ' ...
                        'points should the numbers of function evaluations spent during ' ...
                        'the alorithm execution.'])
                end

                % thresholds should be corresponding to time points when using auc
                if strcmp(Setting.Metric,'auc') && numel(Setting.Thres) ~= numel(Setting.Tmax)
                    error(['The number of thresholds in "Setting.Thres" should be equal to ' ...
                        'the number of time points in "Setting.Tmax". "Setting.Thres" ' ...
                        'refers to the lowest acceptable performance of the design' ...
                        ' algorithms, the performance can be the solution quality.'])
                end
        end
        varargout{1} = Setting;
end
end

File: src/autooptlib/utils/general/input.py
"""Translation of MATLAB Utilities/General/Input.m."""

from __future__ import annotations

import warnings
from types import SimpleNamespace
from typing import Any, Iterable, Sequence


_PARAM_KEYS = {
    "AlgP",
    "AlgQ",
    "Archive",
    "LSRange",
    "IncRate",
    "ProbN",
    "ProbFE",
    "InnerFE",
    "AlgN",
    "AlgFE",
    "AlgRuns",
    "Metric",
    "Compare",
    "Evaluate",
    "Tmax",
    "Thres",
    "RacingK",
    "Surro",
    "AlgFile",
    "AlgName",
}


def _to_sequence(value: Any) -> Sequence[Any]:
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes)):
        return value
    return (value,)


def _ensure_namespace(setting: Any) -> SimpleNamespace:
    if isinstance(setting, SimpleNamespace):
        return setting
    if isinstance(setting, dict):
        return SimpleNamespace(**setting)
    data = {
        name: getattr(setting, name)
        for name in dir(setting)
        if not name.startswith("__") and not callable(getattr(setting, name))
    }
    return SimpleNamespace(**data)


def _find_argument(arguments: Sequence[Any], name: str) -> tuple[bool, Any]:
    if not arguments:
        return False, None
    try:
        idx = arguments.index(name)
    except ValueError:
        return False, None
    if idx + 1 >= len(arguments):
        raise ValueError(f'Missing value for argument "{name}"')
    return True, arguments[idx + 1]


def _to_list(obj: Any) -> list[Any]:
    if isinstance(obj, list):
        return obj
    if isinstance(obj, tuple):
        return list(obj)
    return [obj]


def input_handler(arguments: Iterable[Any], setting: Any, mode: str):
    """Reimplementation of the MATLAB Input.m logic."""
    args = list(arguments)
    set_ns = _ensure_namespace(setting)

    if mode == "data":
        if "Problem" not in args:
            raise ValueError("Please set the targeted problem.")
        _, problem = _find_argument(args, "Problem")
        if set_ns.Mode == "design":
            if "InstanceTrain" not in args or "InstanceTest" not in args:
                raise ValueError("Please set the targeted problem instance indexes.")
            _, train = _find_argument(args, "InstanceTrain")
            _, test = _find_argument(args, "InstanceTest")
            return problem, _to_list(train), _to_list(test)

        if set_ns.Mode == "solve":
            if "InstanceSolve" not in args:
                raise ValueError("Please set the targeted problem instance indexes.")
            _, solve = _find_argument(args, "InstanceSolve")
            return problem, _to_list(solve)

        raise ValueError('Please set the mode to "design" or "solve".')

    if mode == "parameter":
        for key in _PARAM_KEYS:
            found, value = _find_argument(args, key)
            if found:
                setattr(set_ns, key, value)
        return set_ns

    if mode == "check":
        _check_setting(set_ns)
        return set_ns

    raise ValueError(f"Unsupported mode: {mode}")


def _check_setting(setting: SimpleNamespace) -> None:
    mode = getattr(setting, "Mode", None)
    if mode not in {"design", "solve"}:
        raise ValueError('Please set the mode to "design" or "solve".')

    if mode == "design":
        if getattr(setting, "AlgP", 0) > getattr(setting, "AlgQ", 0):
            raise ValueError("The number of pathways should not be larger than the number of operators.")
        if getattr(setting, "AlgN", 0) > getattr(setting, "AlgFE", 0):
            raise ValueError("The number of algorithms should not be larger than the evaluation budget.")
        if getattr(setting, "AlgRuns", 0) > getattr(setting, "ProbFE", 0):
            raise ValueError("The number of runs should not exceed problem evaluations.")

        evaluate = getattr(setting, "Evaluate", "exact")
        compare = getattr(setting, "Compare", "average")
        if evaluate == "racing" and compare != "statistic":
            raise ValueError(
                'The "racing" evaluation method should be used with the algorithm comparing method of "statistic".'
            )
        if evaluate == "racing" and not getattr(setting, "RacingK", None):
            raise ValueError(
                'Please set "Setting.K" as the number of instances evaluated before the first round of racing.'
            )
        if evaluate == "approximate" and not getattr(setting, "Surro", None):
            raise ValueError(
                'Please set "Setting.Surro" as the number of exact performance evaluations when using surrogate.'
            )
        if evaluate == "approximate" and compare == "statistic":
            raise ValueError(
                'It is not necessary to use the "statistic" algorithm comparing method when using the "approximate" evaluation method.'
            )
        if compare == "statistic" and getattr(setting, "AlgRuns", 1) == 1:
            raise ValueError(
                'Please run the design multiple times (Setting.AlgRuns>1) when using the "statistic" comparsion method.'
            )
        if getattr(setting, "ProbN", 0) < 5 and getattr(setting, "AlgP", 0) > 1:
            warnings.warn("It is better to have a large population size if involving the EDA operator", stacklevel=2)
        if getattr(setting, "AlgQ", 0) > 4:
            warnings.warn(
                "AlgQ is recommended to be larger than 4 for discrete and permutation problems due to the lack of so many search operators",
                stacklevel=2,
            )
        if compare == "statistic":
            warnings.warn(
                "It is better to have a large number of training intrances or have a large number of algorithm runs "
                "(set AlgRun to a large number), in order to make the statistical test discriminative.",
                stacklevel=2,
            )

    elif mode == "solve":
        if not getattr(setting, "AlgFile", None) and not getattr(setting, "AlgName", None):
            raise ValueError(
                "Please specify an algorithm file in Setting.AlgFile or specify an algorithm name in Setting.AlgName."
            )

        metric = getattr(setting, "Metric", "quality")
        if metric == "runtimeFE":
            if getattr(setting, "Tmax", None) in (None, []):
                setting.Tmax = getattr(setting, "ProbFE", None)
            if getattr(setting, "Thres", None) in (None, []):
                raise ValueError(
                    'Please set "Setting.Thres" as the lowest acceptable performance of the design algorithms, '
                    "the performance can be the solution quality."
                )
        if metric == "runtimeSec":
            if getattr(setting, "Tmax", None) in (None, []):
                raise ValueError('Please set "Setting.Tmax" as the maximum runtime (seconds).')
            if getattr(setting, "Thres", None) in (None, []):
                raise ValueError(
                    'Please set "Setting.Thres" as the lowest acceptable performance of the design algorithms, '
                    "the performance can be the solution quality."
                )
        if metric == "auc":
            tmax = getattr(setting, "Tmax", None)
            thres = getattr(setting, "Thres", None)
            if not isinstance(tmax, Sequence) or len(tmax) <= 1:
                raise ValueError(
                    '"Setting.Tmax" should contain multiple time points. The time points should the numbers of '
                    "function evaluations spent during the alorithm execution."
                )
            if not isinstance(thres, Sequence) or len(thres) != len(tmax):
                raise ValueError(
                    'The number of thresholds in "Setting.Thres" should be equal to the number of time points in '
                    '"Setting.Tmax". "Setting.Thres" refers to the lowest acceptable performance of the design '
                    "algorithms, the performance can be the solution quality."
                )
